target/cheri-common/cheri_tagmem.c

```diff
@@ -433,7 +564,14 @@ void cheri_tag_set(CPUArchState *env, target_ulong vaddr, int reg, hwaddr* ret_p
     qemu_maybe_log_instr_extra(env, "    Cap Tag Write [" TARGET_FMT_lx "/"
         RAM_ADDR_FMT "] %d -> 1\n", vaddr, ram_offset,
         tag_bit_get(ram_offset >> CAP_TAG_SHFT, ram));
-    tag_bit_set(ram_offset >> CAP_TAG_SHFT, ram);
+    if (tag_bit_set(ram_offset >> CAP_TAG_SHFT, ram)) {
+        // new tag block allocated, flush the TCG tlb so that the magic zero
+        // value is removed from the iotlb.
+        // warn_report("Allocated new tag block for " TARGET_FMT_plx  "->
+        // flushing tlb\r", vaddr);
+        // TODO: assert that the value was the magic zero constant
```

Comment:
Unless I'm missing something, this code is run whenever a tag is set, not just when we allocate a new tag block? So we'll be invalidating the TLB on every valid capability store?

Reply:
It should only return true when a new block is allocated, so in the common case it won't flush.

Reply:
Oh, I see. I feel that might be better as an optional out parameter...

Reply:
Done in new version.